### PR TITLE
[llama4] store expert weights such that we can transpose before grouped mm to have col-major memory layout

### PR DESCRIPTION
**Pull request by danielvegamyhre**
# Summary
Rather than store experts weights pre-transposed (E, in_dim, out_dim), we should store the expert weights non-transposed (E, out_dim, in_dim) then transpose before grouped gemm for (1) compatible dims for gemm, and (2) column-major memory layout required for right operand in grouped gemm. Doing this simple transpose (metadata change only) is must more efficient than doing this [inefficient memory layout transformation before every GEMM in fp8](https://github.com/pytorch/ao/blob/6e941c87c4d9fb9a74e6f979dd522605c696ca42/torchao/prototype/moe_training/scaled_grouped_mm.py#L96). 

# Eager Performance

Llama4 debug model with FSDP=8, using config:
```python
    "debugmodel": TransformerModelArgs(
        dim=5120,
        n_layers=4,
        n_heads=40,
        n_kv_heads=8,
        ffn_dim_multiplier=1.2,
        multiple_of=2048,
        rope_theta=500000,
        max_seq_len=10485760,
        num_experts=16,
        interleave_moe_layer_step=1,
    ),
```

### bfloat16

With change:
```
=====================================================
 Calculating training performance metrics
=====================================================
Median Tokens/Second (excluding step 1): 2147.0
Max Memory Usage: 92.67 GiB
```
Without change:
```
=====================================================
 Calculating training performance metrics
=====================================================
Median Tokens/Second (excluding step 1): 1711.0
Max Memory Usage: 92.67 GiB
```

### fp8 rowwise
With change:
```
(torchtitan) [danvm@devgpu007.eag6 ~/ao/benchmarks/float8/training (metdata)]$ TORCHTITAN_ROOT=/home/danvm/torchtitan NGPU=8 EXTRA_ARGS="--model.converters="float8" --float8.recipe_name="rowwise" --float8.filter_fqns="output,auto_filter_small_kn" --float8.moe_fqns_prototype="experts"" ./llama4.sh

=====================================================
 Calculating training performance metrics
=====================================================
Median Tokens/Second (excluding step 1): 2675.0
Max Memory Usage: 90.35 GiB
```

Without change:
```
(torchtitan) [danvm@devgpu007.eag6 ~/ao/benchmarks/float8/training (metdata)]$ TORCHTITAN_ROOT=/home/danvm/torchtitan NGPU=8 EXTRA_ARGS="--model.converters="float8" --float8.recipe_name="rowwise" --float8.filter_fqns="output,auto_filter_small_kn" --float8.moe_fqns_prototype="experts"" ./llama4.sh

=====================================================
 Calculating training performance metrics
=====================================================
Median Tokens/Second (excluding step 1): 2360.0
Max Memory Usage: 90.35 GiB
```

### CODE REVIEW

File: torchtitan/experiments/llama4/infra/expert_parallel.py [hunk 1]
```diff
@@ -223,18 +226,21 @@ def _token_dispatch(self, mod, inputs, device_mesh):
         return super()._token_dispatch(mod, inputs, self.ep_mesh)
 
     def _partition_fn_2d(self, name, mod, ep_tp_mesh):
+        # w1 shape = (experts, out_dim, in_dim)
         mod.register_parameter(
             "w1",
-            nn.Parameter(distribute_tensor(mod.w1, ep_tp_mesh, [Shard(0), Shard(2)])),
-        )  # Column-wise sharding
```
Comment (tianyu-l) [inline torchtitan/experiments/llama4/infra/expert_parallel.py:229]: let's not remove these comments on colwise vs. rowwise

File: torchtitan/experiments/llama4/model/moe.py
Comment (tianyu-l) [inline torchtitan/experiments/llama4/model/moe.py:1]: please fix files in `torchtitan/models/deepseek_v3` as well

File: torchtitan/experiments/llama4/model/moe.py [hunk 2]
```diff
@@ -23,9 +23,9 @@ def __init__(
...
         self.num_experts = num_experts
-        self.w1 = nn.Parameter(torch.empty(num_experts, dim, hidden_dim))
-        self.w2 = nn.Parameter(torch.empty(num_experts, hidden_dim, dim))
-        self.w3 = nn.Parameter(torch.empty(num_experts, dim, hidden_dim))
+        self.w1 = nn.Parameter(torch.empty(num_experts, hidden_dim, dim))
+        self.w2 = nn.Parameter(torch.empty(num_experts, dim, hidden_dim))
+        self.w3 = nn.Parameter(torch.empty(num_experts, hidden_dim, dim))
         self.use_grouped_mm = use_grouped_mm
 
     def forward(
```
Comment (tianyu-l) [inline torchtitan/experiments/llama4/model/moe.py:31]: we should fix for-loop as well right? Can we do `transpose` in this function, not in the subfunctions?

File: torchtitan/experiments/llama4/model/moe.py
Comment (danielvegamyhre) [reply torchtitan/experiments/llama4/model/moe.py:1]: will do, follow up PR ok?

File: torchtitan/experiments/llama4/model/moe.py (hunk 2 above)
Comment (danielvegamyhre) [reply torchtitan/experiments/llama4/model/moe.py:31]: Yep, this wasn't ready for review yet, still was working on that and running manual tests. It's updated now and I confirmed via setting `use_grouped_mm=False` the loop implementation runs without error.

**Comment by danielvegamyhre** [comment]
cc @tianyu-l for review. In a follow up PR I will update the dsv3 implementation and manually test the changes

File: torchtitan/experiments/llama4/infra/expert_parallel.py [hunk 3]
```diff
@@ -54,17 +54,22 @@ def set_token_group_alignment_size_m(
 # implementation of Tensor Parallel for the GroupedExperts in MoE
 class TensorParallel(ParallelStyle):
     def _partition_fn(self, name, module, device_mesh):
+        # w1 shape = (experts, out_dim, in_dim)
         module.register_parameter(
-            "w1", nn.Parameter(distribute_tensor(module.w1, device_mesh, [Shard(2)]))
-        )  # Column-wise sharding
+            "w1", nn.Parameter(distribute_tensor(module.w1, device_mesh, [Shard(1)]))
+        )  # Rowwise sharding
```
Comment (tianyu-l) [inline torchtitan/experiments/llama4/infra/expert_parallel.py]: reversed

File: torchtitan/experiments/llama4/infra/expert_parallel.py [hunk 4]
```diff
@@ -54,17 +54,22 @@ def set_token_group_alignment_size_m(
...
+
+        # w2 shape = (experts, in_dim, out_dim)
         module.register_parameter(
             "w2",
-            nn.Parameter(distribute_tensor(module.w2, device_mesh, [Shard(1)])),
-        )  # Row-wise sharding
+            nn.Parameter(distribute_tensor(module.w2, device_mesh, [Shard(2)])),
+        )  # Columnwise sharding
```
Comment (tianyu-l) [inline torchtitan/experiments/llama4/infra/expert_parallel.py]: reversed

File: torchtitan/experiments/llama4/infra/expert_parallel.py [hunk 5]
```diff
@@ -223,18 +228,23 @@ def _token_dispatch(self, mod, inputs, device_mesh):
         return super()._token_dispatch(mod, inputs, self.ep_mesh)
 
     def _partition_fn_2d(self, name, mod, ep_tp_mesh):
+        # w1 shape = (experts, out_dim, in_dim)
         mod.register_parameter(
             "w1",
-            nn.Parameter(distribute_tensor(mod.w1, ep_tp_mesh, [Shard(0), Shard(2)])),
-        )  # Column-wise sharding
+            nn.Parameter(distribute_tensor(mod.w1, ep_tp_mesh, [Shard(0), Shard(1)])),
+        )  # Rowwise sharding
```
Comment (tianyu-l) [inline torchtitan/experiments/llama4/infra/expert_parallel.py]: reversed

File: torchtitan/experiments/llama4/infra/expert_parallel.py [hunk 6]
```diff
@@ -223,18 +228,23 @@ def _token_dispatch(self, mod, inputs, device_mesh):
...
+
+        # w2 shape = (experts, in_dim, out_dim)
         mod.register_parameter(
             "w2",
-            nn.Parameter(distribute_tensor(mod.w2, ep_tp_mesh, [Shard(0), Shard(1)])),
-        )  # Row-wise sharding
+            nn.Parameter(distribute_tensor(mod.w2, ep_tp_mesh, [Shard(0), Shard(2)])),
+        )  # Columnwise sharding
```
Comment (tianyu-l) [inline torchtitan/experiments/llama4/infra/expert_parallel.py]: reversed

File: torchtitan/experiments/llama4/model/moe.py [hunk 7]
```diff
@@ -69,9 +69,9 @@ def _run_experts_for_loop(
             )
             out_experts_splits = []
             for expert_idx, x_expert in enumerate(x):
-                h = F.silu(torch.matmul(x_expert, w1[expert_idx]))
-                h = h * torch.matmul(x_expert, w3[expert_idx])
-                h = torch.matmul(h, w2[expert_idx])
+                h = F.silu(torch.matmul(x_expert, w1[expert_idx].transpose(-2, -1)))
+                h = h * torch.matmul(x_expert, w3[expert_idx].transpose(-2, -1))
+                h = torch.matmul(h, w2[expert_idx].transpose(-2, -1))
```
Comment (tianyu-l) [inline torchtitan/experiments/llama4/model/moe.py:74]: can we do this one time in `GroupedExperts.forward` and reuse them for both grouped mm and for-loop?

File: torchtitan/experiments/llama4/infra/expert_parallel.py (hunk 3 above)
Comment (danielvegamyhre) [reply torchtitan/experiments/llama4/infra/expert_parallel.py]: For 3d weights, sharding on dim 1 is applying rowwise sharding for each expert, in my mind. So the new comment is deliberately the reverse of the prior comment. Same applies to the other comments. Can you clarify?

File: torchtitan/experiments/llama4/model/moe.py (hunk 7 above)
Comment (danielvegamyhre) [reply torchtitan/experiments/llama4/model/moe.py:74]: We could but IMO that would make the code less clear/readable if the transpose happens far away from where it is actually used

File: torchtitan/experiments/llama4/infra/expert_parallel.py (hunk 3 above)
Comment (tianyu-l) [reply torchtitan/experiments/llama4/infra/expert_parallel.py]: colwise vs. rowwise is referring to logical placement, not depending on the storage & transpose combination of actual tensors. In TP, you'd first do `intermediate = input x colwise` then `output = intermediate x rowwise`.

As reference, original TP API were designed this way too
https://github.com/pytorch/pytorch/blob/main/torch/distributed/tensor/parallel/style.py#L125

Also you are changing some but not others, e.g. `w3` is still colwise. We should revert all of them. Feel free to add extra comments to explain motivation.

File: torchtitan/experiments/llama4/model/moe.py (hunk 7 above)
Comment (tianyu-l) [reply torchtitan/experiments/llama4/model/moe.py:74]: yeah good point

**Review by tianyu-l — APPROVED** [review]
Please revert the colwise / rowwise comments before landing. See inline comments for details.

File: torchtitan/experiments/llama4/infra/expert_parallel.py (hunk 3 above)
Comment (danielvegamyhre) [reply torchtitan/experiments/llama4/infra/expert_parallel.py]: >colwise vs. rowwise is referring to logical placement, not depending on the storage & transpose combination of actual tensors. 

I see, that makes sense then. Updated the comments accordingly.
